explore/realspace.py

```diff
@@ -603,16 +606,14 @@ def magnetic_sld(qx, qy, up_theta, up_phi, rho, rho_m):
     M = rho_m
     p_hat = np.array([sin_theta * cos_phi, sin_theta * sin_phi, cos_theta ])
 
-    
     q_hat = np.array([qx, qy, 0]) * q_norm
     M_perp = orth(M,q_hat)
     M_perpP = orth(M_perp, p_hat)
     M_perpP_perpQ = orth(M_perpP, q_hat)
 
-    perpx = np.dot(p_hat, M_perp)
+    perpx = p_hat @ M_perp
     perpy = np.sqrt(np.sum(M_perpP_perpQ**2, axis=0))
-    perpz = np.dot(q_hat, M_perpP)
-    
+    perpz = q_hat @ M_perpP
```

Comment:
It turned out that this is overly complicated...
Simpler is:
    p_hat = np.array([sin_theta * cos_phi, sin_theta * sin_phi, cos_theta ])
    #two unit vectors spanning up the plane perpendicular to polarisation for SF scattering (avoiding repetitive computation of orthogonal vectors)
    perpy_hat = np.array([-sin_phi, cos_phi, 0 ])
    perpz_hat = np.array([-cos_theta * cos_phi, -cos_theta * sin_phi, sin_theta ])  

    q_hat = np.array([qx, qy, 0]) * q_norm
    M_perp = orth(M, q_hat)

#then the final results are

    perpx = p_hat @ M_perp
    perpy = perpy_hat @ M_perp
    perpz = perpz_hat @ M_perp

Reply:
Reorganized as:
```python
    # Handle q=0 by setting px = py = 0
    # Note: this is different from kernel_iq, which I(0,0) to 0
    q_norm = 1/sqrt(qx**2 + qy**2) if qx != 0. or qy != 0. else 0.
    q_hat = np.array([qx, qy, 0]) * q_norm

    # perpy_hat and perpz_hat are unit vectors spanning up the plane
    # perpendicular to polarisation for SF scattering (avoiding
    # repetitive computation of orthogonal vectors)
    cos_theta, sin_theta = cos(radians(up_theta)), sin(radians(up_theta))
    cos_phi, sin_phi = cos(radians(up_phi)), sin(radians(up_phi))
    p_hat = np.array([sin_theta * cos_phi, sin_theta * sin_phi, cos_theta])
    perpy_hat = np.array([-sin_phi, cos_phi, 0])
    perpz_hat = np.array([-cos_theta * cos_phi, -cos_theta * sin_phi, sin_theta])

    M_perp = orth(rho_m, q_hat)  # M = rho_m
    perpx = p_hat @ M_perp
    perpy = perpy_hat @ M_perp
    perpz = perpz_hat @ M_perp
```

Reply:
Is magnetic scattering at |q| = 0 defined? The problem is that lim |q| → 0 magnetic sld depends on the direction q hat.

Reply:
For $q=0$, one would see the demagnetising field of the sample, such that $q_hat = np.sqrt(np.array([0.5, 0.5, 0]))$  for a disc shaped sample that is very thin along the beam.

Reply:
It should never be zero, except the magnetisation direction is perpendicular. Something between 1/3 and 1 is most reasonable and physical motivated.